### DISCUSSION TITLE
ETA for RBAC? It's available in V1 but not V2?

### DISCUSSION THREAD

**Discussion by irab** [medusajs/medusa#9909] [Feature Requests]
There's docs and support for roles and access control in V1 but not in v2?

With no middleware in V2 there's not a clear way to implement a rbac resource guard as discussed here?
https://docs.medusajs.com/v1/modules/users/backend/rbac

**Comment by erickirt** [comment]
Check out the delivery example to see how different roles are implemented

**Reply by irab** [reply]
Ah thanks, can see it here. Looks like similar approach

https://github.com/medusajs/medusa-eats/blob/ad3fb769d9af0b0704bd6bc4f19e6cd38b3676b3/backend/src/api/middlewares.ts

**Reply by SalahAdDin** [reply]
It would be better to have an official RBAC module!

**Reply by irab** [reply]
It does seem quite simple though
https://github.com/medusajs/examples/blob/main/marketplace/src/api/middlewares.ts

But yeah, literally no docs on the Medusa Framework or Middleware constructs
https://github.com/medusajs/medusa/blob/develop/packages/core/framework/README.md

https://github.com/medusajs/medusa/blob/develop/packages/core/framework/src/http/middlewares/authenticate-middleware.ts

**Reply by SalahAdDin** [reply]
It is not that easy, as the more control you want to have the more complicated it becomes.

Vendure Commerce has it integrated.

**Reply by radoslaw-sz** [reply]
on our side, we tried our best to not fork the Medusa and just create an extension. Here are the results:

https://github.com/user-attachments/assets/296001d4-0c3e-4435-a650-50aeab6bfc23

**Reply by SalahAdDin** [reply]
will you create a new module?

**Reply by radoslaw-sz** [reply]
by module you mean plugin/extension as NPM package?

**Reply by lmo-wt** [reply]
@radoslaw-sz do you plan to publish your extension? That would be really useful for us!

**Reply by radoslaw-sz** [reply]
@lmo-wt  - yes, there is plan to publish this plugin and also other plugins from RSC Labs, but unfortunately, RBAC exactly won't be available under MIT license, but only commercially. For details you can contact via labs@rsoftcon.com or reach me at Discord.

**Reply by SalahAdDin** [reply]
That's a reason for going to another e-commerce open source tool.

**Reply by 0x4res** [reply]
I start new plugin RBAC for v2 if you want to contribute dm me
https://github.com/Mac-Creations/medusa-plugin-rbac

**Reply by SalahAdDin** [reply]
> I start new plugin RBAC for v2 if you want to contribute dm me https://github.com/Mac-Creations/medusa-plugin-rbac

No code yet.

**Reply by 0x4res** [reply]
> > I start new plugin RBAC for v2 if you want to contribute dm me https://github.com/Mac-Creations/medusa-plugin-rbac
> 
> No code yet.

Up I've already started

**Reply by SalahAdDin** [reply]
> > > I start new plugin RBAC for v2 if you want to contribute dm me https://github.com/Mac-Creations/medusa-plugin-rbac
> > 
> > 
> > No code yet.
> 
> Up I've already started

Let me know if you need help.

**Reply by 0x4res** [reply]
> > > > I start new plugin RBAC for v2 if you want to contribute dm me https://github.com/Mac-Creations/medusa-plugin-rbac
> > > 
> > > 
> > > No code yet.
> > 
> > 
> > Up I've already started
> 
> Let me know if you need help.

The RBAC module is coming soon I have written CRUD for Role, Permission, Member and the UI setting for the admin panel.
The last step is to write the middleware.

Don't hesitate to create PR to write one test for example and I wrote all other tests. I still don't know how the test engine works yet.
And for any suggestions to optimize the code.

**Reply by 0x4res** [reply]
I have a question: is it possible to override the jwt token or add a custom attribute to the token?
I'd like to add the permissions directly in the token to optimize the middleware and avoid a call to the backend.
Thanks

**Reply by SalahAdDin** [reply]
> I have a question: is it possible to override the jwt token or add a custom attribute to the token? I'd like to add the permissions directly in the token to optimize the middleware and avoid a call to the backend. Thanks

It is not a good practice.

**Reply by 0x4res** [reply]
What do you suggest ?

**Reply by irab** [reply]
> I have a question: is it possible to override the jwt token or add a custom attribute to the token?
> 
> I'd like to add the permissions directly in the token to optimize the middleware and avoid a call to the backend.
> 
> Thanks

JWT tokens are signed and designed to be tamper resistant:

https://datatracker.ietf.org/doc/html/rfc7519

**Reply by 0x4res** [reply]
Yes, the best practice is to implement a claim into the jwt token with the permission of the user because the token is signed by the server.
For example, like this.
```
{
  "actor_id": "user_01JFF0T6B9PPHREHG6SKBZ",
  "actor_type": "user",
  "auth_identity_id": "authid_01JFF0T6AC0KTT4Q4TCAXT4E08",
  "app_metadata": {
    "user_id": "user_01JFF0T6B9PPHREHG6SKBZD1ZT"
    "role":[
              {"/product/*": "READ"},
              {"/rbac/*": "WRITE"},
              {"/user/*": "READ"},
    ]
  },
  "iat": 1740381456,
  "exp": 1740467856
}
```

**Reply by SalahAdDin** [reply]
> Yes, the best practice is to implement a claim into the jwt token with the permission of the user because the token is signed by the server. For example, like this.
> 
> ```
> {
>   "actor_id": "user_01JFF0T6B9PPHREHG6SKBZ",
>   "actor_type": "user",
>   "auth_identity_id": "authid_01JFF0T6AC0KTT4Q4TCAXT4E08",
>   "app_metadata": {
>     "user_id": "user_01JFF0T6B9PPHREHG6SKBZD1ZT"
>     "role":[
>               {"/product/*": "READ"},
>               {"/rbac/*": "WRITE"},
>               {"/user/*": "READ"},
>     ]
>   },
>   "iat": 1740381456,
>   "exp": 1740467856
> }
> ```

How is it used on the Admin?

Can we have an architecture like Strapi? with custom hooks and etc.?

**Reply by 0x4res** [reply]
I don’t I search how make this with the emailpass provider but the Token generator of medusja don’t provide this feature
Maybe I do a auth provider

**Reply by SalahAdDin** [reply]
Is it for Entreprise only?
